### PR TITLE
Send integer fields as little endian

### PR DESCRIPTION
**Pull request by keithalucas**
https://github.com/longhorn/longhorn/issues/3875

Signed-off-by: Keith Lucas <keith.lucas@suse.com>

### CODE REVIEW

File: src/longhorn_rpc_protocol.c [hunk 1]
```diff
@@ -45,35 +46,41 @@ static ssize_t write_full(int fd, void *buf, ssize_t len) {
 
 int send_msg(int fd, struct Message *msg) {
         ssize_t n = 0;
+	uint16_t MagicVersion = htole16(MAGIC_VERSION);
+	uint32_t Seq = htole32(msg->Seq);
+	uint32_t Type = htole32(msg->Type);
+	uint64_t Offset = htole64(*((uint64_t *)(&msg->Offset)));
```
Comment (joshimoo) [inline src/longhorn_rpc_protocol.c:52]: Question: I don't remember off the top of my head isn't offset just a int64_t why do we need the conversion to uint64_t pointer to  then deference to get the uint64_t value?

Probably some C magic I don't remember :D

**Review by joshimoo — APPROVED** [review]
LGTM

File: src/longhorn_rpc_protocol.c (hunk 1 above)
Comment (keithalucas) [reply src/longhorn_rpc_protocol.c:52]: So it probably would work if I didn't do this,  my concern is that if we read have a big endian int64_t and want to send it as little endian, the value would often be negative.

As big endian, if the int64_t is stored as bytes `b0 b1 b2 b3 b4 b5 b6 b7`.   It is likely for the most significant bit of the b7 to be 1, i.e. the value will be `0x80-0xFF`.  If I convert this to big endian, the number will be `b7 b6 b5 b4 b3 b2 b1`; b7 will be the most significant byte so the most significant bit of b7 will indicate that the number is negative.

Also, if I look at the implementation of the conversion: 
```#define __bswap_constant_64(x)                  \
  ((((x) & 0xff00000000000000ull) >> 56)        \
   | (((x) & 0x00ff000000000000ull) >> 40)      \
   | (((x) & 0x0000ff0000000000ull) >> 24)      \
   | (((x) & 0x000000ff00000000ull) >> 8)       \
   | (((x) & 0x00000000ff000000ull) << 8)       \
   | (((x) & 0x0000000000ff0000ull) << 24)      \
   | (((x) & 0x000000000000ff00ull) << 40)      \
   | (((x) & 0x00000000000000ffull) << 56))
```

I don't know if x is signed, if bitwise anding it with the unsigned value will make it signed or unsigned, if it is signed and the most significant byte is 0x80-0xFF, the right shift will extend the sign and make it something like 0xFFFFFFFFFFFFxx.  So I did this to completely avoid any possibility of using the sign of the value as `int64_t` in the endianness.

**Review by innobead — COMMENTED** [review]
LGTM

File: src/longhorn_rpc_protocol.c [hunk 2]
```diff
@@ -103,37 +111,57 @@ int receive_msg(int fd, struct Message *msg) {
...
                 errorf("fail to read type\n");
 		return -EINVAL;
         }
-        n = read_full(fd, &msg->Offset, sizeof(msg->Offset));
-        if (n != sizeof(msg->Offset)) {
+
+	msg->Type = le32toh(msg->Type);
+
+        n = read_full(fd, &Offset, sizeof(Offset));
+        if (n != sizeof(Offset)) {
```
Comment (derekbit) [inline src/longhorn_rpc_protocol.c:140]: @keithalucas 
Should it be `msg->Offset` rather than `Offset`?

Comment (keithalucas) [reply src/longhorn_rpc_protocol.c:140]: I read the value into a temporary unsigned variable then convert it to the signed value a little after this.

**Pull request merged**